File: lecture_6/Lecture6. Source code/topic_tutorial_py/direction.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from std_msgs.msg import Bool

class DirectionChecker(Node):

    def __init__(self):
        super().__init__('direction_checker')
        self.subscription = self.create_subscription(
            Twist,
            'turtle1/cmd_vel',
            self.listener_callback,
            10)
        self.publisher_ = self.create_publisher(Bool, 'direction_status', 10)

    def listener_callback(self, msg):
        bool_msg = Bool()
        if msg.linear.x < 0:
            bool_msg.data = False
            self.get_logger().info('Publishing: "False"')
        else:
            bool_msg.data = True
            self.get_logger().info('Publishing: "True"')
        self.publisher_.publish(bool_msg)

def main(args=None):
    rclpy.init(args=args)
    direction_checker = DirectionChecker()
    rclpy.spin(direction_checker)
    direction_checker.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: lecture_6/src/topic_tutorial_cpp/src/direction.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/bool.hpp"
#include "geometry_msgs/msg/twist.hpp"

class DirectionCheck : public rclcpp::Node
{
public:
    DirectionCheck() : Node("check_direction_node")
    {
        subscription_ = this->create_subscription<geometry_msgs::msg::Twist>(
            "turtle1/cmd_vel", 10, std::bind(&DirectionCheck::twistCallback, this, std::placeholders::_1));
        publisher_ = this->create_publisher<std_msgs::msg::Bool>("linear_x_sign", 10);

        RCLCPP_INFO(this->get_logger(), "DirectionCheck node has been started.");
    }
private:
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr subscription_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr publisher_;

    void twistCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
    {
        std_msgs::msg::Bool bool_msg;
        bool_msg.data = (msg->linear.x >= 0);

        if (bool_msg.data)
        {
            RCLCPP_INFO(this->get_logger(), "Bool Linear_x: True");
        }
        else
        {
            RCLCPP_INFO(this->get_logger(), "Bool Linear_x: False");
        }

        publisher_->publish(bool_msg);
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DirectionCheck>();

    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}